rpmlint/checks/FileDigestCheck.py

```diff
@@ -34,46 +38,90 @@ def _get_digest_group_from_path(self, path):
+                result = False
+        return result
 
-            filename = pkgfile.name
+    def _is_valid_digest(self, pkgfile, digest):
+        algorithm = digest['algorithm']
+        if algorithm == 'skip':
+            return True
 
-            if filename in pkg.ghost_files:
-                self.output.add_info('E', pkg, f'{group}-file-digest-ghost', filename)
-            elif known_digests and filename in known_digests:
-                expected_digests = known_digests[filename]
-                if 'skip' in expected_digests:
-                    pass
-                else:
-                    h = hashlib.new('sha256')
-                    with open(pkgfile.path, 'rb') as fd:
-                        while True:
-                            chunk = fd.read(4096)
-                            if not chunk:
-                                break
-                            h.update(chunk)
-                    hexdigest = h.hexdigest()
-                    if hexdigest not in expected_digests:
-                        self.output.add_info('E', pkg, f'{group}-file-digest-mismatch', filename,
-                                             f'expected:{",".join(expected_digests)}', f'has:{hexdigest}')
-            elif group:
-                self.output.add_info('E', pkg, f'{group}-file-digest-unauthorized', filename)
+        digest_hash = digest['hash']
+        pair = (pkgfile.name, algorithm)
+        if pair not in self.digest_cache:
+            h = hashlib.new(algorithm)
+            with open(pkgfile.path, 'rb') as fd:
+                while True:
+                    chunk = fd.read(4096)
+                    if not chunk:
+                        break
+                    h.update(chunk)
+            self.digest_cache[pair] = h.hexdigest()
+
+        return self.digest_cache[pair] == digest_hash
+
+    def _check_group_digests(self, digest_group, pkg):
+        for filename, digest in digest_group.items():
+            if not self._is_valid_digest(pkg.files[filename], digest):
+                return False
+        return True
+
+    def check_binary(self, pkg):
+        """
+        Check that all files in secured locations are covered by a file digest group
+        in which all files have valid digest.
+        """
+
+        if not self._check_symlinks(pkg):
+            return
+
+        # First collect all files that are in a digest configuration group
+        secured_paths = {pkgfile.name for pkgfile in pkg.files.values() if self._in_digest_configuration_group(pkgfile)}
+
+        # Iterate all digest groups and find one that covers all secured files
+        # and in which all digests match
+        for digest_group in self.digest_groups:
+            covered_files = set(digest_group.keys())
+            if secured_paths - covered_files:
+                # bail out as some secured paths are not covered
+                continue
```

Comment:
There is are now missing precise diagnostic messages towards the packager. For example for the `sarg` package case with the related `/usr/bin/sarg-reports` digests we would get three error messages for files in /etc/cron.daily, /etc/cron.weekly, /etc/cron.monthly but no error message at all for /usr/bin/sarg-reports, if the digest for only /usr/bin/sarg-reports would be mismatched.

In our rpmlint1 check we remembered the verification results of the "most recent" whitelisting group and displayed these results as diagnostic towards the packager. So I would like to be able to show the packager:
- file a/b: wrong digest
- file c/d: missing digest
- etc.

The question is how we can now determine what the "most recent" whitelisting entry is. Starting with the highest boo# or bsc# number would probably be a good match. This would also safe us a couple of digest calculations in the "good" verification case, because we don't need to check outdated digests before we encounter the actually matching ones.

Reply:
Hmm, to be honest doing proper diagnostics is quite tricky. I don't think the "most recent" bug is always what you want: in the case of a SLE-produce, I would expect an older bug to be used. All that can be achieved is a hint and I'm curious if worth for the compilation. Moreover, the pass now does not compare digests if it's obvious that a configuration group does not cover all "secured" files.

Reply:
For SLE products we have our dedicated branches where we only backport individual whitelistings so the assumption should typically hold true.

From my experience with working with packagers and my coworkers exactly these bits of information are important and can save an hour or so of trying to figure out what is going on. And it is especially confusing when e.g. an error like described above is shown for /etc/cron.weekly/sarg-something while actually there is a valid digest for it.

Reply:
I agree with mgerstner. We have to consider here that this pops up for users that might have never had any experience with rpmlint. So having error message to be as clear as possible is a high priority in my opinion. And as we had this for a while in rpmlint(1) it would be sad to lose it in the rewrite. I understand the strive for clean code but that is something that would cause pain for us and others over a long time.

Reply:
Be my guest. I improved the error handling and now should be reported a group for which we have a minimal number of errors. That should work as the best hint for rpmlint consumers.